grid/chat/messages/maps/slack.suma

```diff
@@ -0,0 +1,230 @@
+profile = "chat/messages@1.0"
+provider = "slack"
+
+"""
+GetMessages map
+"""
+map GetMessages {
+  http GET "/conversations.history" {
+    request "application/x-www-form-urlencoded" {
+      query {
+        channel = input.destination
+        latest = input.beforeTimestamp
+        oldest = input.afterTimestamp
+        cursor = input.page
+        limit = input.limit || 999
+      }
+
+      headers {
+        "Authorization" = `Bearer ${parameters.accessToken}`
+      }
+    }
+
+    response 200 "application/json" {
+      error = call MapSlackError(
+        statusCode = statusCode,
+        error = body,
+        headers = headers,
+      ) if (!body.ok)
+
+      return map error if (!body.ok) error
+
+      messages = call MapMessages(
+        messages = body.messages,
+        destination = input.destination,
+      )
+
+      rateLimit = call MapRateLimitDetails(headers = headers)
+      result.messages = messages
+
+      set if (body.has_more) {
+        result.nextPage = body.response_metadata.next_cursor
+      }
+
+      set if (rateLimit) {
+        result.rateLimit = rateLimit
+      }
+
+      map result result
+    }
+
+    response {
+      error = call MapSlackError(
+        statusCode = statusCode,
+        error = body,
+        headers = headers,
+      )
+      
+      map error error
+    }
+  }
+}
+
+operation MapMessages {
+  listOfMessages = []
+
+  listOfMessages = args.messages.filter(message => message.type === 'message').map(message => {
+    const resultMessage = {
+      id: message.ts,
+      text: message.text,
+      createdAt: parseFloat(message.ts),
+    }
+
+    let author, reactions = [], attachments = [];
+
+    if (message.user || message.bot_id) {
+      author = {
+        id: message.user || message.bot_id,
+        username: message.username
+      };
+
+      resultMessage.author = author;
+    }
+
+    if (message.thread_ts) {
+      resultMessage.hasThread = true;
+      resultMessage.threadId = message.thread_ts;
+    }
+
+    if (message.edited) {
+      resultMessage.updatedAt = parseFloat(message.edited.ts);
+    }
+
+    if (message.reactions) {
+      reactions = message.reactions.map(reaction => ({
+        emoji: reaction.name,
+        count: reaction.count,
+        users: reaction.users,
+      }));
+
+      resultMessage.reactions = reactions;
+    }
+
+    if (message.files) {
+      attachments = message.files.map(file => ({
+        id: file.id,
+        createdAt: file.created,
+        fileName: file.name,
+        mediaType: file.mimetype,
+        url: file.permalink_public,
+        preview: file.preview,
+      }));
+
+      resultMessage.attachments = attachments;
+    }
+
+    resultMessage.url = `https://slack.com/api/conversations.history?channel=${args.destination}&latest=${message.ts}&inclusive=true&limit=1`;
```

Comment:
Okay, so this is last outstanding issue for me.

I see that Slack has a dedicated method for that, which would be costly to run for every message: https://api.slack.com/methods/chat.getPermalink

But I'd say consumer's expectation here is to return a URL which is presentable to the end user and take them to the relevant message. So returning this is in my opinion worse than returning nothing at all.

What is missing to be able to construct this URL manually. For example here is threaded message:

```
https://superfaceai.slack.com/archives/C013602CL2Y/p1645175948429299?thread_ts=1645043940.215569&cid=C013602CL2Y
```

I don't think we have any workspace identifier (or do we?), but we do have channel ID, message ID and thread ID.

Reply:
I'll try to create public URL then. 🙏🏻

Reply:
I would also need workspace name to compose correct URL, so it's not possible without that call.

We could return partially generated url, user would have to add his workspace name. For example, I'm able to generate URL: `https://slack.com/archives/CF3H7S63W/p1644310634546009`, but to make it work, I have to add workspace name before domain i.e. `https://<workspace-name>.slack.com/archives/CF3H7S63W/p1644310634546009`. URL for threads are more complicated, but still need workspace name.